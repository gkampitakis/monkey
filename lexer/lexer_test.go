package lexer_test

import (
	"fmt"
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/gkampitakis/monkey/lexer"
	"github.com/gkampitakis/monkey/token"
)

func TokensSnapshot(t *testing.T, input string) {
	t.Helper()

	l := lexer.New([]byte(input))
	b := strings.Builder{}
	for {
		tok := l.NextToken()
		b.WriteString(fmt.Sprintf("[%s]=>%q", tok.Type, tok.Literal))
		if tok.Type == token.EOF {
			break
		}
		b.WriteByte('\n')
	}

	snaps.MatchSnapshot(t, b.String())
}

func TestNextToken(t *testing.T) {
	t.Run("simple input", func(t *testing.T) {
		input := `=+(){},;`

		TokensSnapshot(t, input)
	})

	t.Run("monkey source code input", func(t *testing.T) {
		input := `let five = 5;

		let ten = 10;

		let add = fn(x, y) {

			x + y;

		};

		let result = add(five, ten);
		`

		TokensSnapshot(t, input)
	})

	t.Run("monkey source code more identifiers", func(t *testing.T) {
		input := `let fine = 5;

		let ten = 10;

		let add = fn(x, y) {

			x+y;

		};

		let result = add(five, ten);

		!-/*5;

		5 < 10 > 5;
		
		if (5 < 10) {
			return true;
		} else {
			return false;
		}

		10 == 10;
		10 != 9;
		let value = 10+10;
		"foobar"
		"foo bar"
		"hello \n world \t"		
		[1, 2];
		{"foo":"bar"}
		`

		TokensSnapshot(t, input)
	})
}
