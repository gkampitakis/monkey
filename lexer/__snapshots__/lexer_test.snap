
[TestNextToken/simple_input - 1]
[ASSIGN]=>"="
[PLUS]=>"+"
[LPAREN]=>"("
[RPAREN]=>")"
[LBRACE]=>"{"
[RBRACE]=>"}"
[COMMA]=>","
[SEMICOLON]=>";"
[EOF]=>""
---

[TestNextToken/monkey_source_code_input - 1]
[LET]=>"let"
[IDENT]=>"five"
[ASSIGN]=>"="
[INT]=>"5"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"ten"
[ASSIGN]=>"="
[INT]=>"10"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"add"
[ASSIGN]=>"="
[FUNCTION]=>"fn"
[LPAREN]=>"("
[IDENT]=>"x"
[COMMA]=>","
[IDENT]=>"y"
[RPAREN]=>")"
[LBRACE]=>"{"
[IDENT]=>"x"
[PLUS]=>"+"
[IDENT]=>"y"
[SEMICOLON]=>";"
[RBRACE]=>"}"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"result"
[ASSIGN]=>"="
[IDENT]=>"add"
[LPAREN]=>"("
[IDENT]=>"five"
[COMMA]=>","
[IDENT]=>"ten"
[RPAREN]=>")"
[SEMICOLON]=>";"
[EOF]=>""
---

[TestNextToken/monkey_source_code_more_identifiers - 1]
[LET]=>"let"
[IDENT]=>"fine"
[ASSIGN]=>"="
[INT]=>"5"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"ten"
[ASSIGN]=>"="
[INT]=>"10"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"add"
[ASSIGN]=>"="
[FUNCTION]=>"fn"
[LPAREN]=>"("
[IDENT]=>"x"
[COMMA]=>","
[IDENT]=>"y"
[RPAREN]=>")"
[LBRACE]=>"{"
[IDENT]=>"x"
[PLUS]=>"+"
[IDENT]=>"y"
[SEMICOLON]=>";"
[RBRACE]=>"}"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"result"
[ASSIGN]=>"="
[IDENT]=>"add"
[LPAREN]=>"("
[IDENT]=>"five"
[COMMA]=>","
[IDENT]=>"ten"
[RPAREN]=>")"
[SEMICOLON]=>";"
[BANG]=>"!"
[MINUS]=>"-"
[SLASH]=>"/"
[ASTERISK]=>"*"
[INT]=>"5"
[SEMICOLON]=>";"
[INT]=>"5"
[LT]=>"<"
[INT]=>"10"
[GT]=>">"
[INT]=>"5"
[SEMICOLON]=>";"
[IF]=>"if"
[LPAREN]=>"("
[INT]=>"5"
[LT]=>"<"
[INT]=>"10"
[RPAREN]=>")"
[LBRACE]=>"{"
[RETURN]=>"return"
[TRUE]=>"true"
[SEMICOLON]=>";"
[RBRACE]=>"}"
[ELSE]=>"else"
[LBRACE]=>"{"
[RETURN]=>"return"
[FALSE]=>"false"
[SEMICOLON]=>";"
[RBRACE]=>"}"
[INT]=>"10"
[EQ]=>"=="
[INT]=>"10"
[SEMICOLON]=>";"
[INT]=>"10"
[NEQ]=>"!="
[INT]=>"9"
[SEMICOLON]=>";"
[LET]=>"let"
[IDENT]=>"value"
[ASSIGN]=>"="
[INT]=>"10"
[PLUS]=>"+"
[INT]=>"10"
[SEMICOLON]=>";"
[STRING]=>"foobar"
[STRING]=>"foo bar"
[STRING]=>"hello \\n world \\t"
[LBRACKET]=>"["
[INT]=>"1"
[COMMA]=>","
[INT]=>"2"
[RBRACKET]=>"]"
[SEMICOLON]=>";"
[EOF]=>""
---
